test/test_device_radix_sort.cu

```diff
@@ -490,104 +495,214 @@ cudaError_t Dispatch(
+      .doit(CDPDispatchKernel<IsDescending,
+                              CubBackend,
+                              KeyT,
+                              ValueT,
+                              BeginOffsetIteratorT,
+                              EndOffsetIteratorT,
+                              NumItemsT>,
+            is_descending,
+            cub_backend,
+            d_selector,
+            d_temp_storage_bytes,
+            d_cdp_error,
+            d_temp_storage,
+            temp_storage_bytes,
+            d_keys,
+            d_values,
+            num_items,
+            num_segments,
+            d_segment_begin_offsets,
+            d_segment_end_offsets,
+            begin_bit,
+            end_bit,
+            debug_synchronous);
+  };
+
+  // This unreachable branch is intentional -- without this, the device pass
+  // won't see the kernel instantiation and won't compile anything.
+  if (false)
+  {
+    launch_kernel(); // Fake kernel launch for device pass.
+  }
 
-    // Copy out error
-    cudaError_t retval;
-    CubDebugExit(cudaMemcpy(&retval, d_cdp_error, sizeof(cudaError_t) * 1, cudaMemcpyDeviceToHost));
-    return retval;
+  NV_IF_TARGET(
+    NV_IS_HOST,
+    (
+      // Invoke kernel to invoke device-side dispatch. Actual kernel launch:
+      CubDebugExit(launch_kernel());
+      CubDebugExit(cub::detail::device_synchronize());
+
+      // Copy out selector
+      CubDebugExit(cudaMemcpy(&d_keys.selector,
+                              d_selector,
+                              sizeof(int) * 1,
+                              cudaMemcpyDeviceToHost));
+      d_values.selector = d_keys.selector;
+
+      // Copy out temp_storage_bytes
+      CubDebugExit(cudaMemcpy(&temp_storage_bytes,
+                              d_temp_storage_bytes,
+                              sizeof(size_t) * 1,
+                              cudaMemcpyDeviceToHost));
+
+      // Copy out error
+      CubDebugExit(cudaMemcpy(&retval,
+                              d_cdp_error,
+                              sizeof(cudaError_t) * 1,
+                              cudaMemcpyDeviceToHost));
+  )); // NV_IS_HOST
+
+  return retval;
 }
 
+// Specializations of Dispatch that translate the CDP backend to the appropriate
+// CUB backend, and uses the CUB backend to launch the CDP kernel.
+#define DEFINE_CDP_DISPATCHER(CdpBackend, CubBackend)                          \
+  template <int IsDescending,                                                  \
+            typename KeyT,                                                     \
+            typename ValueT,                                                   \
+            typename BeginOffsetIteratorT,                                     \
+            typename EndOffsetIteratorT,                                       \
+            typename NumItemsT>                                                \
+  CUB_RUNTIME_FUNCTION cudaError_t Dispatch(                                   \
```

Comment:
Does this `Dispatch` function and `LaunchCDPKernel` has to be annotated as `CUB_RUNTIME_FUNCTION`? Could they always be `__host__` ones? This would allow us to remove `IF_TARGET` from these functions, right?

Reply:
Oooh, nice suggestion, I hadn't considered that and just matched the signature of the other `Dispatch` overloads. This cleans things up a lot! 👍